staging/src/k8s.io/component-base/featuregate/testing/feature_gate.go

```diff
@@ -67,9 +67,7 @@ func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f fea
 	tb.Cleanup(func() {
 		tb.Helper()
 		detectParallelOverrideCleanup()
-		if err := gate.(featuregate.MutableFeatureGate).Set(fmt.Sprintf("%s=%v", f, originalValue)); err != nil {
-			tb.Errorf("error restoring %s=%v: %v", f, originalValue, err)
-		}
+		gate.(featuregate.MutableVersionedFeatureGateForTests).Reset(originalEnabled)
```

Comment:
This looks like a mistake.  If called from multiple tests (imagine them nested for instance), setting different flags, this would seem to reset everything instead of only the changed featuregate.

Reply:
It would not reset everything. The flags from last step are stored in `originalEnabled`. We don't want to store all the set feature values along the path to explicitly set flags because the following scenario:
1. in test1, emulationVersion = 1.30, `AllAlpha=true`, which results in some alpha feature in 1.30 `featureA= true`. With the old implementation `gate.(featuregate.MutableFeatureGate).Set`, it would explicitly set `featureA=false`
2. in test2, emulationVersion = 1.31, featureA becomes beta feature with default true, but it is explicitly set to `featureA=false` from last step. This is not correct.

Reply:
Please update this function to narrowly reverse the change that is made instead of doing a blanket reset.

Reply:
updated the function. 
Also added additional tests and checking of emulation version does not change